File: ws-python/tests/integration_test.py
# Copyright 2026 TEAM RelicFrog
# SPDX-License-Identifier: Apache-2.0
"""Integration tests for primes_cli."""

from __future__ import annotations

import pytest

from primes_cli.primes import (
    is_prime,
    nth_prime,
    primes_in_range,
    sieve_of_eratosthenes,
)


class TestNthPrimeConsistentWithSieve:
    def test_first_50_primes(self) -> None:
        sieve = sieve_of_eratosthenes(230)  # 230 > 50th prime (229)
        for i, expected in enumerate(sieve, start=1):
            assert nth_prime(i) == expected, f"nth_prime({i}) should be {expected}"


class TestPrimesInRangeFirstHundred:
    def test_matches_sieve(self) -> None:
        sieve = sieve_of_eratosthenes(541)  # 541 == 100th prime
        rang = primes_in_range(2, 541)
        assert len(rang) == 100
        assert rang == sieve


class TestIsPrimeConsistentWithSieve:
    def test_up_to_200(self) -> None:
        sieve_set = set(sieve_of_eratosthenes(200))
        for n in range(201):
            assert is_prime(n) == (n in sieve_set), f"is_prime({n}) disagrees with sieve"


class TestErrorPropagation:
    def test_sieve_limit_zero(self) -> None:
        with pytest.raises(ValueError):
            sieve_of_eratosthenes(0)

    def test_range_inverted(self) -> None:
        with pytest.raises(ValueError):
            primes_in_range(50, 10)

    def test_nth_prime_zero(self) -> None:
        with pytest.raises(ValueError):
            nth_prime(0)


class TestLargePrimes:
    @pytest.mark.parametrize("p", [7919, 104729, 999_983])
    def test_known_large_primes(self, p: int) -> None:
        assert is_prime(p) is True

    @pytest.mark.parametrize("c", [7920, 104728, 999_999])
    def test_known_large_composites(self, c: int) -> None:
        assert is_prime(c) is False

File: ws-zig/src/integration_test.zig
// Copyright 2026 TEAM RelicFrog
// SPDX-License-Identifier: Apache-2.0
//
// tests/integration.zig — Integration tests for primes-cli (Zig).

const std = @import("std");
const primes = @import("primes.zig");

test "round-trip: nthPrime consistent with sieve (first 50)" {
    const alloc = std.testing.allocator;
    const sieve = try primes.sieveOfEratosthenes(alloc, 230);
    defer alloc.free(sieve);

    for (sieve, 0..) |expected, i| {
        const got = try primes.nthPrime(@intCast(i + 1));
        try std.testing.expectEqual(expected, got);
    }
}

test "primesInRange first 100 matches sieve" {
    const alloc = std.testing.allocator;
    const sieve = try primes.sieveOfEratosthenes(alloc, 541);
    defer alloc.free(sieve);

    const range = try primes.primesInRange(alloc, 2, 541);
    defer alloc.free(range);

    try std.testing.expectEqual(@as(usize, 100), range.len);
    try std.testing.expectEqualSlices(u64, sieve, range);
}

test "isPrime consistent with sieve for 0..200" {
    const alloc = std.testing.allocator;
    const sieve = try primes.sieveOfEratosthenes(alloc, 200);
    defer alloc.free(sieve);

    var sieve_set = std.AutoHashMap(u64, void).init(alloc);
    defer sieve_set.deinit();
    for (sieve) |p| try sieve_set.put(p, {});

    var n: u64 = 0;
    while (n <= 200) : (n += 1) {
        const got = primes.isPrime(n);
        const want = sieve_set.contains(n);
        try std.testing.expectEqual(want, got);
    }
}

test "error propagation: sieve limit=0" {
    const alloc = std.testing.allocator;
    try std.testing.expectError(primes.PrimeError.LimitTooSmall, primes.sieveOfEratosthenes(alloc, 0));
}

test "error propagation: range start > end" {
    const alloc = std.testing.allocator;
    try std.testing.expectError(primes.PrimeError.InvalidRange, primes.primesInRange(alloc, 50, 10));
}

test "error propagation: nth 0" {
    try std.testing.expectError(primes.PrimeError.InvalidNthIndex, primes.nthPrime(0));
}

test "comptime table sanity: all 1000 entries are prime" {
    for (primes.PRIMES_TABLE) |p| {
        try std.testing.expect(primes.isPrime(p));
    }
}

test "large primes detected" {
    for ([_]u64{ 7919, 104729, 999_983 }) |p| {
        try std.testing.expect(primes.isPrime(p));
    }
}

test "large composites rejected" {
    for ([_]u64{ 7920, 104728, 999_999 }) |c| {
        try std.testing.expect(!primes.isPrime(c));
    }
}
